Page Templates: Track template insertion/removal caused by the template selector

marekhrabe opened the pull request:
#### Changes proposed in this Pull Request

* adds a filter in SPT so the tracking library can detect events being triggered during template insertion

#### Testing instructions

No need to checkout this PR. Just apply the diff below. 

* apply D38786-code to your sandbox (diff just for testing purposes, real deployment would be handled separately after merging this PR)
* pick a wpcom simple site with SPT active for testing
* sandbox that site.wordpress.com and widgets.wp.com
* on your (sandboxed) site, go to the editor and run `localStorage.debug = '*'` (it's easier to go through wp-admin for testing than calypso)
* refresh the page
* start a new page, you should see the layout selector and `wpcom-block-editor:tracking registering tracking handlers` message in console (you may need to use devtools' `filter` feature to filter your console logs by the term `track`)
* pick a layout
* watch dev console for messages, one of them should be `wpcom-block-editor:analytics:tracks Recording event "wpcom_block_inserted"` and in its props, you should see `from_template_selector: true`
* insert any other block manually to the page and there should be one more tracking message but *without* `from_template_selector`

Fixes #39357

marekhrabe committed:
experimental implementation using global var

marekhrabe committed:
experimental implementation using global var

marekhrabe committed:
switch global variable to filter

marekhrabe committed:
switch global variable to filter

marekhrabe committed:
explain filter

marekhrabe committed:
explain filter

retrofox commented on apps/wpcom-block-editor/src/wpcom/features/tracking.js:
```diff
@@ -73,11 +76,30 @@ const trackBlockReplacement = ( originalBlockIds, blocks ) => {
+			block_name: block.name,
+			blocks_replaced: false,
+			// isInsertingPageTemplate filter is set by Starter Page Templates
+			from_template_selector: applyFilters( 'isInsertingPageTemplate', false ),
```
Wondering if does it worth defining a global constants file in order to share them across these plugins.

obenland commented:
This looks pretty solid already. What are the next steps?

marekhrabe commented:
> This looks pretty solid already. What are the next steps?

Testing instructions now added 🙌

getdave commented:
Going to review this today.

getdave commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:
```diff
@@ -431,6 +445,9 @@ export const PageTemplatesPlugin = compose(
 			},
 			insertTemplate: ( title, blocks ) => {
+				// Add filter to let the tracking library know we are inserting a template.
+				addFilter( INSERTING_HOOK_NAME, INSERTING_HOOK_NAMESPACE, stubTrue );
```
With `stubTrue` being the default value?

marekhrabe replied:
yeah, we need it to always return true and this is a convenience method for just that. we remove the filter after insertion so it will be back to `false`

getdave commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:
```diff
@@ -2,7 +2,18 @@
+	filter,
+	sortBy,
+	stubTrue,
+} from 'lodash';
```
These reformatting changes make it difficult to see what was added. However, given it was just `stubTrue` that's ok 👍 😄

marekhrabe replied:
only `stubTrue` ✅

getdave commented on apps/wpcom-block-editor/src/wpcom/features/tracking.js:
```diff
@@ -73,11 +76,30 @@ const trackBlockReplacement = ( originalBlockIds, blocks ) => {
+			block_name: block.name,
+			blocks_replaced: false,
+			// isInsertingPageTemplate filter is set by Starter Page Templates
+			from_template_selector: applyFilters( 'isInsertingPageTemplate', false ),
```
So by default, this is `false` right? It's odd because when I inserted a Columns block the tracks event was missing `from_template_selector` prop.

<img width="1328" alt="Screenshot 2020-02-13 at 11 35 57" src="https://user-images.githubusercontent.com/444434/74430755-71ffd600-4e55-11ea-8303-7430f2196c2c.png">

But then once the Block inserted it automatically added x2 `core/column` (singular!) Blocks and those did have the `from_template_selector` prop:

<img width="932" alt="Screenshot 2020-02-13 at 11 40 01" src="https://user-images.githubusercontent.com/444434/74430838-978cdf80-4e55-11ea-8445-7227f4088a0f.png">

getdave replied:
Also...just checking, if a filter isn't registered then I assume anything that hooks into that filter is ignored and doesn't throw an error. Just checking because otherwise if you don't ensure widgets is deployed in advance of this PR landing we're going to have errors.

marekhrabe replied:
>  if a filter isn't registered then I assume anything that hooks into that filter is ignored

100% correct. it just returns the default value immediately, which is `false`

marekhrabe replied:
> Columns block the tracks event was missing from_template_selector prop.

When you insert a block manually, it uses a different tracking code path where `from_template_selector` can never be true so that's why the check is not included.

> Block inserted it automatically added x2 core/column (singular!) Blocks and those did have the from_template_selector prop:

It's using the same method template selector uses (`replaceInnerBlocks`) so the tracking will include `from_template_selector` and it should be `false` because no filter is registered at the moment to change it to anything else

getdave replied:
Thanks for thorough explanations.

getdave commented on apps/wpcom-block-editor/src/wpcom/features/tracking.js:
```diff
@@ -1,8 +1,10 @@
  */
 import { use, select } from '@wordpress/data';
 import { registerPlugin } from '@wordpress/plugins';
+import { applyFilters } from '@wordpress/hooks';
```
Using a filter is really nice 👍

marekhrabe replied:
very WordPress!

getdave reviewed (dismissed):
This is looking really good.

I tested as per your (quite excellent) instructions and it all seemed to work perfectly.

The only query I have is around why when I manually add Blocks the `from_template_selector` prop appears to be either not present or set to `false`. Surely it should always be set to `false` no?

Other than that I'd say this is good to go (with the exception of getting ✅  on your tests!)

retrofox commented:
LGTM. Tested in my sandbox and tracks when the blocks are inserted from the template selector:

![image](https://user-images.githubusercontent.com/77539/74722039-0764e680-5217-11ea-9aed-cd5702046557.png)

question: it doesn't track blocks contained into other blocks. For instance, the `about` template...
![image](https://user-images.githubusercontent.com/77539/74722234-5874da80-5217-11ea-8feb-585f6e018aa3.png)

...has many blocks which are contained by the `core/group block`. Are we going to ignore tracking these blocks? Make sense of tracking them?

In fact, on this specific example, when the template is inserted we are tracking that only `core/group` blocks are been inserted, but the template is full of other ones such as image, paragraph, columns, etc

getdave commented:
<img width="1676" alt="Screenshot 2020-02-18 at 10 35 24" src="https://user-images.githubusercontent.com/444434/74728317-6c7e0380-523a-11ea-810b-c50424c3349c.png">

Same concern here @retrofox. I see only `core/group` events being tracked. I'm going to go through code again to get a fuller understanding and see if I can use this to inform what is going on.

marekhrabe commented:
You are correct, I was about to make an issue for recursive tracking as I have noticed it too with my testing. I would say it's out of scope for this change, rather a new feature that we need to plan for iteration. Would you agree?

retrofox commented:
> You are correct, I was about to make an issue for recursive tracking as I have noticed it too with my testing. I would say it's out of scope for this change, rather a new feature that we need to plan for iteration. Would you agree?

Honestly, to me, it's like a blocker since the implementation doesn't tackle the main purpose of the PR. 
Also, populating partially the stats registering only `core/group` events and ignoring the other blocks seem to be something not very well.
Anyway, it's totally up to you. We can continue iterating immediately after merging this one. 😃👍

getdave commented:
> Honestly, to me, it's like a blocker since the implementation doesn't tackle the main purpose of the PR

I am on the fence here. Initially I agreed with @retrofox but now remember my recent troubles, I'm inclined towards merging what we have and following up with incremental improvements.

Tracking something might be better than nothing unless it significantly skews the event data to make it look like `core/group` is the best block ever (which it is of course! 😉).

getdave commented on apps/wpcom-block-editor/src/wpcom/features/tracking.js:
```diff
@@ -73,11 +76,30 @@ const trackBlockReplacement = ( originalBlockIds, blocks ) => {
+ * @returns {void}
+ */
+const trackInnerBlocksReplacement = ( rootClientId, blocks ) => {
+	castArray( blocks ).forEach( block => {
```
If before `castArray()` `blocks` is a deeply nested object then presumably we can recurse into it and call the same callback you provide to `forEach` to handle dispatching events for _all_ blocks and not just the top level?

getdave replied:
Just checked and `blocks` is

> Block objects to insert as new InnerBlocks

To presumably it is nested.

marekhrabe replied:
We can flatten the blocks here as you suggest. I haven't done it immediately as I was wondering if other actions we are tracking should also be looked at recursively, hence the suggestion to make it a separate task and get answers first.

retrofox replied:
>  I was wondering if other actions we are tracking should also be looked at recursively

Yes, it's a very good point, too.

retrofox committed:
wpcom-block-editor: traking tpl blocks recursively
it tracks all blocks recursively (and asynchronously) when are inserted by the template, one by one, with a delay of 50ms in order to avoid send multiple events at the same time.

retrofox committed:
wpcom-block-editor: traking tpl blocks recursively
it tracks all blocks recursively (and asynchronously) when are inserted by the template, one by one, with a delay of 50ms in order to avoid send multiple events at the same time.

getdave commented on apps/wpcom-block-editor/src/wpcom/features/tracking.js:
```diff
@@ -84,14 +84,34 @@ const trackBlockReplacement = ( originalBlockIds, blocks ) => {
+		blockCounter++;
+
+		if ( block.innerBlocks && block.innerBlocks.length ) {
+			trackInnerBlocksReplacement( block.clientId, block.innerBlocks );
```
Yep this is the recursion that we need 👍

retrofox committed:
wpcom-block-editor: refactoring adding trackBlocks()

retrofox committed:
wpcom-block-editor: refactoring adding trackBlocks()

getdave reviewed (commented):
I tested this with @retrofox's changes to add recursion and it appears to be working in that we now have a large number of Blocks being tracked whereas previously it was only tracking the top-level `core/group`.

I feel like this is an improvement over the original as we are now actually tracking all the blocks that get inserted and not just the topmost ones.

We should probably be careful about taking this any further. We should consider shipping "as is" and following up with further refinements in future PRs.

Great work @marekhrabe and @retrofox!

### Testing Details
<img width="1864" alt="Screenshot 2020-02-20 at 11 54 05" src="https://user-images.githubusercontent.com/444434/74931678-08914180-53d8-11ea-9425-816fcd1a719d.png">

I chose this template for testing:

<img width="1923" alt="Screenshot 2020-02-20 at 11 57 05" src="https://user-images.githubusercontent.com/444434/74931725-23fc4c80-53d8-11ea-8012-5837763edf04.png">

getdave commented on apps/wpcom-block-editor/src/wpcom/features/tracking.js:
```diff
@@ -29,6 +29,57 @@ const getTypeForBlockId = blockId => {
+			},
+			trackerBlocksCounter * 50,
+			block,
+			parentBlock
```
I see these are args to the callback function within the timeout right?

retrofox replied:
yeah, it ensures that we are using the correct var values in the function scope.

retrofox commented:
I've updated added some changes in order to track all blocks, and it populates the event with `inner_block`, `parent_block_client_id` and `parent_block_name`

![image](https://user-images.githubusercontent.com/77539/74932256-20f56200-53c0-11ea-866f-3b524f8b3a97.png)

getdave commented on apps/wpcom-block-editor/src/wpcom/features/tracking.js:
```diff
@@ -84,14 +84,34 @@ const trackBlockReplacement = ( originalBlockIds, blocks ) => {
+	}
+
+	castBlocks.forEach( block => {
+		setTimeout(
```
Maybe you are trying to defer the calls to avoid blocking the editor from rendering? If so then this will push each call onto the next "tick" of the Event loop:

```js
setTimeout(fn, 0)
```

Guess it's much the same. Either way we should comment this to make sure it's clear what's going on.

> This method can be used to break up long running operations and run a callback function immediately after the browser has completed other operations such as events and display updates.

retrofox replied:
Yeah.
First of all, probably it doesn't make sense.

One reason to delay tracking these events is avoiding to affect with the template rendering process. A template could contain hundreds of blocks and waiting to track all events could affect it in terms of performance. Again, not sure about this.
My other concern is triggering these requests at the same time. For this reason, I've delayed, one by one, every tacking event. I think that 200 requests at the same time could be something undesired, but definitely it worths research and confirm this fragile theory.
Anyway, removing this will get the approach easier and simpler. Glad to do it.

getdave commented on apps/wpcom-block-editor/src/wpcom/features/tracking.js:
```diff
@@ -26,6 +29,57 @@ const getTypeForBlockId = blockId => {
+				if ( _parent ) {
+					eventProperties.parent_block_client_id = _parent.clientId;
+					eventProperties.parent_block_name = _parent.name;
+				}
```
I was wondering about this. I think it's good to have this info.

retrofox replied:
yeah, it's related whit the @marekhrabe comment

>  ... haven't done it immediately as I was wondering if other actions we are tracking should also be looked at recursively,...

marekhrabe replied:
I'm not sure the client_id would be useful. These ids are an implementation detail of the block editor and they are not persistent across edit sessions so I don't think we could gain anything from having them. The block name though is a perfect idea!

marekhrabe commented on apps/wpcom-block-editor/src/wpcom/features/tracking.js:
```diff
@@ -73,11 +127,28 @@ const trackBlockReplacement = ( originalBlockIds, blocks ) => {
+const trackInnerBlocksReplacement = ( rootClientId, blocks ) => {
+	trackBlocks( blocks, 'wpcom_block_inserted', block => ( {
+		block_name: block.name,
+		blocks_replaced: false,
```
I think this should be true when doing a block replacement.

retrofox replied:
does it mean that `blocks_replaced` should be true only for the root blocks? So for the inner blocks it should be false?

marekhrabe replied:
If the newly inserted block replaced anything, it should be true. I think inner blocks should inherit it, as they are part of the replacement.

marekhrabe commented:
The functionality looks good but it doesn't feel right with the `setTimeout` and a module-global variable added. It feels strange to solve the concurrency problem on this level. Did you encounter any serious perf issues without it or did it just felt wrong in general to do all of those requests at once? 

If it's not already critical, I would suggest removing the `setTimeout` from here and calling `tracksRecordEvent` immediately in the loop. If we encounter performance problems, I would update the `tracksRecordEvent` function with some limitation logic so all tracking calls would get the benefit, not just this single loop.

retrofox commented:
> The functionality looks good but it doesn't feel right with the `setTimeout` and a module-global variable added. It feels strange to solve the concurrency problem on this level.

Yeah, far enough.

> Did you encounter any serious perf issues without it or did it just felt wrong in general to do all of those requests at once?

No, I don't. This implementation is based on a simple assumption of mine.


> If it's not already critical, I would suggest removing the `setTimeout` from here and calling `tracksRecordEvent` immediately in the loop. If we encounter performance problems, I would update the `tracksRecordEvent` function with some limitation logic so all tracking calls would get the benefit, not just this single loop.

👍
Thanks

retrofox committed:
wpcom-block-editor: remove async handling

retrofox committed:
wpcom-block-editor: remove async handling

retrofox committed:
wpcom-block-editor: improve jsdoc

retrofox committed:
wpcom-block-editor: improve jsdoc

retrofox committed:
wcom-editor-blocks: trick inner blocks when insert

marekhrabe commented:
I think this is ready now, awaiting one last approval after all new changes introduced.

For the shipping, we need to publish a new release of the Full Site Editing Plugin PCYsg-mMA-p2. After that is done (the order is important), we need to update `wpcom-block-editor` as per PCYsg-l4k-p2

retrofox committed:
wpcom-block-editor: set blocks_replaced to true when inserted

retrofox committed:
wpcom-block-editor: set blocks_replaced to true when inserted

obenland commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:
```diff
@@ -24,6 +36,8 @@ import ensureAssets from './utils/ensure-assets';
 
 const DEFAULT_HOMEPAGE_TEMPLATE = 'maywood';
+const INSERTING_HOOK_NAME = 'isInsertingPageTemplate';
+const INSERTING_HOOK_NAMESPACE = 'automattic/full-site-editing/inserting-template';
```
Should we define these in a separate file so they can be imported in `tracking.js`?

retrofox replied:
I made the same suggestion at the beginning, but in some way, it's like in WordPress filters, right? You need to know the exact name of the filter in order to get rightly hooked.
Anyway, I agree with adding a global file definition, too.

marekhrabe replied:
Since FSE and wpcom-block-editor are completely separate apps and I believe there is no precedent of any shared code other than npm-based libraries, I think we should keep them separate. In WordPress backend you also write all the filter names directly in files and you don't use a central place with a list of filters to import.

obenland commented:
@marekhrabe It looks like that phab diff is out of date. Was there a way to test the tracking locally by setting a variable in the console?

getdave commented:
Testing now

getdave reviewed (approved):
I tested this on WPCom using WPAdmin. 

I selected the "About" template:

<img width="1271" alt="Screenshot 2020-02-25 at 10 51 43" src="https://user-images.githubusercontent.com/444434/75240824-e3784680-57bc-11ea-83d2-0084f69bc43a.png">

...and I saw these events in console when filtering by the term `track`:

<img width="1264" alt="Screenshot 2020-02-25 at 10 51 12" src="https://user-images.githubusercontent.com/444434/75240845-ee32db80-57bc-11ea-9826-1caaeab0bc10.png">

These appear to marry up well with that is in the Template. I believe the recursive tracking is working.

getdave commented on apps/wpcom-block-editor/src/wpcom/features/tracking.js:
```diff
@@ -26,6 +29,49 @@ const getTypeForBlockId = blockId => {
 
+/**
+ * This helper function tracks the given blocks recursively
+ * in order to be able to do it also for its inner ones.
```
```suggestion
 * in order to track inner blocks.
```

getdave commented on apps/wpcom-block-editor/src/wpcom/features/tracking.js:
```diff
@@ -26,6 +29,49 @@ const getTypeForBlockId = blockId => {
+ * the parent block (if exists). Take this as
+ * an advantage to add other custom properties to the event.
+ *
+ * Also, it adds as default `inner_block`,
```
```suggestion
 * Also, it adds default `inner_block`,
```

getdave commented on apps/wpcom-block-editor/src/wpcom/features/tracking.js:
```diff
@@ -65,19 +109,34 @@ const trackBlockInsertion = blocks => {
+
+/**
+ * Track inner blocks replacement.
+ * This is how Page Templates insert their content into page, by replacing everything that was already there.
```
```suggestion
 * Page Templates insert their content into the page replacing everything that was already there.
```

getdave reviewed (commented):
Some tiny minor comment wording tweaks I forgot on my other review.

marekhrabe committed:
Update apps/wpcom-block-editor/src/wpcom/features/tracking.js

Co-Authored-By: Dave Smith <getdavemail@gmail.com>

marekhrabe committed:
Update apps/wpcom-block-editor/src/wpcom/features/tracking.js

Co-Authored-By: Dave Smith <getdavemail@gmail.com>

marekhrabe committed:
Update apps/wpcom-block-editor/src/wpcom/features/tracking.js

Co-Authored-By: Dave Smith <getdavemail@gmail.com>

marekhrabe commented:
Accepted all of them, thanks a lot :)

marekhrabe committed:
fix code after rebase
